### ISSUE TITLE
Fresh matplotlib install broken: installs qt5, but matplotlib rc patched to require qt4

### ISSUE THREAD

**Issue by jorisvandenbossche** [ContinuumIO/anaconda-issues#1068]
Related to https://github.com/ContinuumIO/anaconda-issues/issues/1065 (but opening a new issue as this is not a Qt problem but how the matplotlib rc is patched AFAIU). Also reported it at matplotlib: https://github.com/matplotlib/matplotlib/issues/7124#issuecomment-247762925

When you do a fresh `conda install matplotlib`, this now installs qt/pyqt 5.x instead of 4.x. 
But as far as I understand, anaconda patches matplotlib to default to qt4 backend. With the consequence that you get a `ImportError: No module named 'PyQt4'` error if you do not specify the backend to qt5agg manually.

Possible workarounds for now (either one of the two):

- pin pyqt to 4.11
- set the backend manually before importing pyplot (`matplotlib.use('qt5agg')`)

**Comment by jorisvandenbossche** [comment]
The patch is here: https://github.com/ContinuumIO/anaconda-recipes/blob/master/matplotlib/rctmp_pyside.patch

**Comment by ccordoba12** [comment]
Thanks @jorisvandenbossche, this means that we need to remove or update that patch so that Matplotlib can work with Qt5.

**Comment by tacaswell** [comment]
You can also fix this by setting the backend in any of the ways documented at http://matplotlib.org/faq/usage_faq.html#what-is-a-backend to `'Qt5Agg'`

**Comment by cbrnr** [comment]
@ccordoba12 is there a PR that fixes this issue?

**Comment by ccordoba12** [comment]
Not yet, sorry.

**Comment by astrofrog** [comment]
Just to give a heads-up, this is causing builds to fail for Astropy and also for Glue - so we are pinning qt=4 for now.

**Comment by ccordoba12** [comment]
Thanks @astrofrog, we're still fighting with this ;-)

**Comment by michaelaye** [comment]
Just as an addition, this is not only happening for anaconda, but a simple update in my conda environment created this issue as well.

**Comment by ccordoba12** [comment]
This should be fixed now. @astrofrog, could you confirm it?

**Comment by michaelaye** [comment]
by what version of files? I did the update today and got this issue.

**Comment by jorisvandenbossche** [comment]
Just did a fresh test of `conda create -n test-mpl python=3.5 matplotlib -c defaults`, still get the error

**Comment by tacaswell** [comment]
I am also not convinced that having a hard dependency on qt5 in the right solution.  This is going to conflict badly with anyone who is embedding matplotlib in a qt4 app and is directly import pyqt4 rather than going through some sort of compatibility library.

**Comment by ccordoba12** [comment]
@tacaswell, what do you propose instead?

**Comment by tacaswell** [comment]
Probably something in the install script of pyqt and/or mpl that sets default when the system knows which version of qt is installed.

**Comment by tacaswell** [comment]
or provide a build with each version and let the resolver sort it out?

**Comment by astrofrog** [comment]
Is there any reason the Matplotlib build even needs to require PyQt as a dependency? I often get frustrated that the whole of Qt gets installed when I really just want to use the Agg backend. Can't we decouple Matplotlib from PyQt, and just rely on the normal default Matplotlib behavior?

**Comment by ilanschnell** [comment]
To me is seems the problem is that the `~/.matplotlib` configuration is still mentioning `PyQt4`.  If this is the case, the easiest way to solve this problem is to remove the matplotlib configuration, such that the new matplotlib 1.5.3 can create a new one (which will refer to `PtQy5`), as https://github.com/ContinuumIO/anaconda-recipes/blob/master/matplotlib/rctmp_pyside.patch is updated.

**Comment by tacaswell** [comment]
As part of the build process mpl tries to guess what the current GUI toolkit is an set that to be the 'default' backend, else 'Agg'.  If you do not ship a patch `.config/matplotlibrc` than you will get piles of bug reports that mpl is not working because it is set to a non-interactive backend.

Probably the right way to package this is to have a `matplotlib-core` with no gui deps, `matplotlib-qt4` which depends on qt4 + mpl-core and ships a properly patched config file for pyqt4, a `matplotlib-qt5` package which does the same for qt5 and then leave `matplotlib` as an alias for which ever flavor you prefer.

This scheme would also allow gtk / wx toolkits to be supported in the future.

**Comment by jbednar** [comment]
+1 for a matplotlib-core that non-GUI packages using matplotlib can depend on!  That would really make our lives much easier.

**Comment by ccordoba12** [comment]
There are some discussions already to support this kind of packages in conda-build. But until then, we need to fix this :-)

**Comment by rcthomas** [comment]
The second workaround isn't working for me (fresh Linux Anaconda 3.5 installation).  Anyone else having trouble using either work-around?

    > python --version
    Python 3.5.2 :: Anaconda 4.2.0 (64-bit)
    > cat try.py
    import matplotlib
    matplotlib.use("qt5agg")
    import matplotlib.pyplot as plt
    > python try.py
    Traceback (most recent call last):
      File "try.py", line 3, in <module>
        import matplotlib.pyplot as plt
      File "/global/homes/r/rthomas/tmp/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py", line 114, in <module>
        _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
      File "/global/homes/r/rthomas/tmp/anaconda3/lib/python3.5/site-packages/matplotlib/backends/__init__.py", line 32, in pylab_setup
        globals(),locals(),[backend_name],0)
      File "/global/homes/r/rthomas/tmp/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt5agg.py", line 16, in <module>
        from .backend_qt5 import QtCore
      File "/global/homes/r/rthomas/tmp/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt5.py", line 31, in <module>
        from .qt_compat import QtCore, QtGui, QtWidgets, _getSaveFileName, __version__
      File "/global/homes/r/rthomas/tmp/anaconda3/lib/python3.5/site-packages/matplotlib/backends/qt_compat.py", line 137, in <module>
        from PyQt4 import QtCore, QtGui
    ImportError: No module named 'PyQt4'
    > conda info
    Current conda install:

                   platform : linux-64
              conda version : 4.2.9
           conda is private : False
          conda-env version : 4.2.9
        conda-build version : 2.0.2
             python version : 3.5.2.final.0
           requests version : 2.11.1
           root environment : /global/homes/r/rthomas/tmp/anaconda3  (writable)
        default environment : /global/homes/r/rthomas/tmp/anaconda3
           envs directories : /global/homes/r/rthomas/tmp/anaconda3/envs
              package cache : /global/homes/r/rthomas/tmp/anaconda3/pkgs
               channel URLs : https://repo.continuum.io/pkgs/free/linux-64/
                              https://repo.continuum.io/pkgs/free/noarch/
                              https://repo.continuum.io/pkgs/pro/linux-64/
                              https://repo.continuum.io/pkgs/pro/noarch/
                config file : /global/homes/r/rthomas/.condarc
               offline mode : False

**Comment by tskisner** [comment]
I see it too from a fresh Miniconda3 install with only numpy, matplotlib, and dependencies installed.  I already purged ~/.matplotlib but no change.

**Comment by tacaswell** [comment]
What version of qt, pyqt, and mpl do you have installed?

What is the output of `matplotlib.get_backend()` ?

**Comment by rcthomas** [comment]
@tacaswell, in my case:

    > conda list | grep qt
    pyqt                      5.6.0                    py35_0    defaults
    qt                        5.6.0                         0    defaults
    qtawesome                 0.3.3                    py35_0    defaults
    qtconsole                 4.2.1                    py35_1    defaults
    qtpy                      1.1.2                    py35_0    defaults
    > conda list | grep matplotlib
    matplotlib                1.5.3               np111py35_0    defaults
    > python -c "import matplotlib; print(matplotlib.get_backend())"
    Qt5Agg

Hmmm!

**Comment by tskisner** [comment]
@rcthomas and I are both using the same Cray system at NERSC.  I think this might be a local problem that we can solve offline.  My versions are:
```
$ conda list | grep qt
pyqt                      5.6.0                    py35_0  
qt                        5.6.0                         1  
$ conda list | grep matplotlib
matplotlib                1.5.3               np111py35_0  
$ python -c "import matplotlib; print(matplotlib.get_backend())"
Qt5Agg 
```
I get the above results on both the login and compute nodes.  If I run a script that actually tries to make a plot then the backend imports pyqt.  On the login node this works fine.  On a compute node I get:
```
import matplotlib.pyplot as plt                                                                         
  File "/global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/python3.5/site-packages/matplotlib/pyplot.py", line 114, in <module>                                                                   
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()                            
  File "/global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/python3.5/site-packages/matplotlib/backends/__init__.py", line 32, in pylab_setup
    globals(),locals(),[backend_name],0)
  File "/global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/python3.5/site-packages/matplotlib/backends/backend_qt5agg.py", line 16, in <module>
    from .backend_qt5 import QtCore
  File "/global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/python3.5/site-packages/matplotlib/backends/backend_qt5.py", line 31, in <module>
    from .qt_compat import QtCore, QtGui, QtWidgets, _getSaveFileName, __version__
  File "/global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/python3.5/site-packages/matplotlib/backends/qt_compat.py", line 137, in <module>
    from PyQt4 import QtCore, QtGui
ImportError: No module named 'PyQt4'
```
For some reason qt_compat.py is loading PyQt4 instead of 5.  So I think this problem is likely on our end.

**Comment by tskisner** [comment]
Hi there again.  Please disregard my previous post- this problem is reproducible on all nodes of this system.  I have dug a little deeper now...

The system is running SLES 11 sp3.  From a fresh, bare-bones miniconda3 install (just numpy, matplotlib, and dependencies with versions above), matplotlib attempts to load pyqt5 and fails, then tries to fall back to pyqt4 (which does not exist).  So this problem lies with pyqt5.

Looking into pyqt5:
```
$ python
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.

>>> from PyQt5 import QtCore, QtGui, QtWidgets
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /usr/lib64/libxcb-xlib.so.0: undefined symbol: _xcb_unlock_io
```
This is strange, since the conda qt5 package is built against the conda libxcb.  Yet the system library is being picked up.  Looking more at the qt5 X11 bindings (asterisks added for emphasis):
```
$ ldd /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/libQt5X11Extras.so                                                                                                         
        linux-vdso.so.1 =>  (0x00002aaaaaaab000)                                                            
        libQt5Gui.so.5 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/./libQt5Gui.so.5 (0x00002aaaaacb1000)                                                                                
        libQt5Core.so.5 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/./libQt5Core.so.5 (0x00002aaaab504000)                                                                              
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00002aaaabd06000)                                      
        libGL.so.1 => /usr/lib64/libGL.so.1 (0x00002aaaabf23000)                                            
        libstdc++.so.6 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/./libstdc++.so.6 (0x00002aaaac1aa000)                                                                                
        libm.so.6 => /lib64/libm.so.6 (0x00002aaaac566000)                                                  
        libgcc_s.so.1 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/./libgcc_s.so.1 (0x00002aaaac7df000)                                                                                  
        libc.so.6 => /lib64/libc.so.6 (0x00002aaaac9f7000)                                                  
        libpng16.so.16 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/././libpng16.so.16 (0x00002aaaacd73000)                                                                              
        libz.so.1 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/././libz.so.1 (0x00002aaaacfb5000)                                                                                        
        libicui18n.so.54 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/././libicui18n.so.54 (0x00002aaaad1cb000)                                                                          
        libicuuc.so.54 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/././libicuuc.so.54 (0x00002aaaad630000)                                                                              
        libicudata.so.54 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/././libicudata.so.54 (0x00002aaaad9ce000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00002aaaaf3f8000)
        libgthread-2.0.so.0 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/././libgthread-2.0.so.0 (0x00002aaaaf5fd000)
        librt.so.1 => /lib64/librt.so.1 (0x00002aaaaf7fe000)
        libglib-2.0.so.0 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/././libglib-2.0.so.0 (0x00002aaaafa07000)
        /lib64/ld-linux-x86-64.so.2 (0x0000555555554000)
        libglapi.so.0 => /usr/lib64/libglapi.so.0 (0x00002aaaafd46000)
        libXdamage.so.1 => /usr/lib64/libXdamage.so.1 (0x00002aaaaff9d000)
        libXfixes.so.3 => /usr/lib64/libXfixes.so.3 (0x00002aaab01a0000)
        libX11-xcb.so.1 => /usr/lib64/libX11-xcb.so.1 (0x00002aaab03a7000)
**      libxcb-glx.so.0 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/./libxcb-glx.so.0 (0x00002aaab05a9000)
        libXxf86vm.so.1 => /usr/lib64/libXxf86vm.so.1 (0x00002aaab07c1000)
        libXext.so.6 => /usr/lib64/libXext.so.6 (0x00002aaab09c8000)
        libX11.so.6 => /usr/lib64/libX11.so.6 (0x00002aaab0bda000)
**      libxcb-xlib.so.0 => /usr/lib64/libxcb-xlib.so.0 (0x00002aaab0f17000)
**      libxcb.so.1 => /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/./libxcb.so.1 (0x00002aaab111a000)
        libXau.so.6 => /usr/lib64/libXau.so.6 (0x00002aaab133f000)
        libdrm.so.2 => /usr/lib64/libdrm.so.2 (0x00002aaab1543000)
        libXdmcp.so.6 => /usr/lib64/libXdmcp.so.6 (0x00002aaab1750000)
```
So you can see it is picking up libxcb and libxcb-glx from the conda package, but libxcb-xlib from the ancient system version.  Some searching shows that modern libxcb does not even include libxcb-xlib.so, and that library is not included in the conda package.

If the qt5 conda package is built against the libxcb conda package, then it seems like the dependency on libxcb-xlib should not be there.  Maybe the qt5 build is finding/using system libraries on the build machine?

**Comment by tskisner** [comment]
I should mention that the workaround of pinning pyqt to 4.11 **does** work for me to fix the matplotlib error.  I'll open a separate issue related to "Qt5 installs on ancient SLES systems" if/when that becomes important.

**Comment by tacaswell** [comment]
@rcthomas What does `import PyQt5` do?

**Comment by rcthomas** [comment]
@tacaswell, to answer your question, and again with a completely fresh installation (Anaconda3-4.2.0-Linux-x86_64.sh):

    > python
    Python 3.5.2 |Anaconda 4.2.0 (64-bit)| (default, Jul  2 2016, 17:53:06)
    [GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import PyQt5
    >>>

Now I was able to make this all just go away by doing `conda uninstall libxcb` --- I expected this not to work but it appears it doesn't uninstall anything but Anaconda libxcb.  Then everything just falls back to the system X11 stuff and I can do interactive plotting again.

@tskisner was not able to do this with his Miniconda setup.

**Comment by tskisner** [comment]
FYI, the import of the top-level pyqt will not trigger the loading of the qt shared libraries.  but doing:
```
from PyQt5 import QtCore, QtGui, QtWidgets
```
will.  @rcthomas - after doing your "conda uninstall libxcb", can you confirm the backend that is being used again?

**Comment by rcthomas** [comment]
After `conda uninstall libxcb`

    > python -c "import matplotlib; print(matplotlib.get_backend())"
    Qt5Agg

**Comment by dmacmillan** [comment]
@rcthomas I am experiencing the same issue as you are. I cannot simply set the backend and get matplotlib to work. Also @tskisner I tried running your command for importing qt X11 shared libraries and got this:

`python -c "from PyQt5 import QtCore, QtGui, QtWidgets"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: libXdmcp.so.6: cannot open shared object file: No such file or directory`

**Comment by ccordoba12** [comment]
@rcthomas, please install that library (libXdmcp.so.6) with your package manager :-)

**Comment by ccordoba12** [comment]
@tskisner, how ancient is your system? We compile our packages in CentOS 5 :-)

**Comment by dmacmillan** [comment]
Sigh, I don't have a package manager on my system. I need to do everything manually because of restrictions. I looked into installing this library but it requires another library (xproto) which didn't work or I couldn't figure out how to do a local install even with `./configure --prefix=/custom/path`

**Comment by mingwandroid** [comment]
@tskisner: Can you show the output of:
`ldd -v /global/common/edison/contrib/hpcosmo/conda/toast-conda_3.5-20161011/lib/libQt5X11Extras.so`

please?

**Comment by jjhelmus** [comment]
The linkage to libXdmcp seems to be coming from the libxcb package (1.12-0), specifically the libxcb.so libarary in that package.  

```
(test) [root@ac1c8b82b8d0 /]# ldd /opt/conda/envs/test/lib/libxcb.so
	linux-vdso.so.1 =>  (0x00007ffcfb1ed000)
	libXau.so.6 => /usr/lib64/libXau.so.6 (0x00007f920511e000)
	libXdmcp.so.6 => not found
	libc.so.6 => /lib64/libc.so.6 (0x00007f9204d89000)
	/lib64/ld-linux-x86-64.so.2 (0x000055920cf99000)
```

At least on the system on am testing on (CentOS6) force uninstalling the libxcb package reverts to using the system provided libxcb library which do not link against libXdmcp.

**Comment by rcthomas** [comment]
@ccordoba12, @tskisner and I are running on a Cray XC30.  Update via package manager is not really our call here.  There is a planned OS upgrade coming but it won't be for several months.

**Comment by tskisner** [comment]
@mingwandroid , the ldd output was in my post further up this thread...  Since this seems to be a Qt5 + libxcb + system conflict problem (not matplotlib or pyqt5), perhaps we should open a separate issue?

**Comment by mingwandroid** [comment]
Can you try:

```
echo "" |  gcc -shared -x c -o libXdmcp.so.6 -
LD_PRELOAD=./libXdmcp.so.6 python
from PyQt5 import QtCore, QtGui, QtWidgets
```

**Comment by mingwandroid** [comment]
@tskinser: I asked for `-v` to be added, but it's ok, I've investigated further now and don't need it now.

**Comment by tskisner** [comment]
@mingwandroid sorry for missing that '-v'...  The preload of the empty shared library has no effect.  Just to be clear:  @dmacmillan is seeing a different system library problem (no libXdmcp.so on the system).

On the system that @rcthomas and I are using, that library exists and seems to have no unresolved symbols.  Our issue instead is with the libxcb-xlib.so library, which exists in older versions of libxcb, but no longer exists in newer upstream versions of the package.

I think I have tracked down the library dependency that is the problem:

1.  libQt5X11Extras depends on libGL (from /usr) and on libxcb (from conda).

2.  libGL also depends on libxcb and various libxcb-* libraries

3.  On newer systems, all the libxcb-* system libraries needed by libGL are also provided by the conda package, and all is well.

4.  On some older systems, libGL also depends on libxcb-xlib, provided by the ancient libxcb version in the OS.  Now when libQt5X11Extras loads libGL, the modern libxcb from conda does not provide libxcb-xlib, and so libGL finds the system one.

Obviously this is the slippery slope faced by all userspace package management.  Eventually you end up replacing the whole user space :-)

**Comment by jjhelmus** [comment]
@mingwandroid 

From a CentOS 6 based docker container without libXdmcp installed in a environment created using `conda create -n test python=2.7 pyqt`:

```
(test) [root@8fe68af61bad /]# echo "" |  gcc -shared -x c -o libXdmcp.so.6 -
(test) [root@8fe68af61bad /]# LD_PRELOAD=./libXdmcp.so.6 python
Python 2.7.12 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:42:40)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> from PyQt5 import QtCore, QtGui, QtWidgets
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: libXdmcp.so.6: cannot open shared object file: No such file or directory
```

**Comment by ccordoba12** [comment]
So we have two problems here:

1. Our `libxcb` is linking to `libXdmcp`, which is not present in some platforms. According to [Linux from Scratch](http://www.linuxfromscratch.org/blfs/view/svn/x/libxcb.html), `libXdmcp` is an optional dependency, so we could look at compiling `libxcb` without it.
2. Our `libxcb` does not provide `libxcb-xlib`, needed by some platforms. There's no solution to this problem, as far as I can tell :-) Qt 5 [requires](http://doc.qt.io/qt-5/linux-requirements.html) `libxcb` **1.8.1** or higher, and it seems the latest version to contain `libxcb-xlib` was **1.1**, at least according to [this link](http://changelog.debian.net/libxcb) (look for *libxcb-xlib is no more* :-)

**Comment by mingwandroid** [comment]
@jjhelmus, can you try to update `libxcb` as follows:

`conda update -c qttesting libxcb`

and let me know if that fixes your issue please?

For the `libxcb-xlib` problem, I think I need to fire up a SLES 11 sp3 VM and investigate further.

**Comment by jjhelmus** [comment]
@mingwandroid The `libxcb` in `qttesting` seems to solve the `libXdmcp` issue:

```
(test) [root@bcc31561ee0f /]# python -c "from PyQt5 import QtCore, QtGui, QtWidgets"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: libXdmcp.so.6: cannot open shared object file: No such file or directory
(test) [root@bcc31561ee0f /]# conda install -c qttesting libxcb
Fetching package metadata .........
Solving package specifications: ..........

Package plan for installation in environment /opt/conda/envs/test:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libxcb-1.12                |                1         1.5 MB  qttesting

The following packages will be UPDATED:

    libxcb: 1.12-0 defaults --> 1.12-1 qttesting

Proceed ([y]/n)? y

Fetching packages ...
libxcb-1.12-1. 100% |#####################################################################| Time: 0:00:02 516.98 kB/s
Extracting packages ...
[      COMPLETE      ]|########################################################################################| 100%
Unlinking packages ...
[      COMPLETE      ]|########################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|########################################################################################| 100%
(test) [root@bcc31561ee0f /]# python -c "from PyQt5 import QtCore, QtGui, QtWidgets"
(test) [root@bcc31561ee0f /]#
```

Importing matplotlib.pyplot in this environment also works.

**Comment by mingwandroid** [comment]
Thanks for the feedback @jjhelmus, I just need to figure out what to do about libxcb-xlib now, I'm fairly stumped on that one. Dropping OpenGL is one possibility, but that's drastic (though CentOS 5.11 has pretty limited OpenGL support anyway ...)

I'll upload the new `libxcb` to `defaults` anyway.

**Comment by David-OConnor** [comment]
I can confirm that on a clean install of Anaconda py3.5 on win64, this is still present. Steps to reproduce:

-Install Anaconda
-conda update conda; conda update --all
-from matplotlib import pyplot as plt

You can revert to an old version of qt to allow matplotlib to work, but then jupyter qtconsole doesn't work.

**Comment by mingwandroid** [comment]
@David-OConnor: I cannot reproduce this. The matplotlib 1.5.3 package contains a `Lib\site-packages\matplotlib\mpl-data\matplotlibrc` file which contains the following:

```
backend      : Q5Agg
```

If you can try it again and do `conda list` I'd appreciate it, one thing to note is that matplotlib has a list of places it looks in for matplotlibrc, as detailed here: http://stackoverflow.com/a/23484838/3257826

.. so it could be pretty site-specific which of those got used.

I am leaning towards closing this issue as I think it's morphed into a load of different issues:

- @tacaswell's wish for split packages (feature request, conda-build has some WIP for that)
- The SELS11-is-too-old issue (won't fix)
- @jjhelmus' libxcb issue (fixed)


If people wish, they can create new, more specific ones.

**Comment by David-OConnor** [comment]
Turns out the issue was with a line in my ipython config called '%matplotlib qt'; removing this line fixes it.

**Comment by mingwandroid** [comment]
OK, thanks @David-OConnor.

**Comment by jorisvandenbossche** [comment]
Since a simple install of matplotlib in a clean environment still results in a broken import, I would personally not close this issue.

**Comment by mingwandroid** [comment]
It doesn't though, I've not been able to repeat that on Linux, Windows nor macOS.

**Comment by jorisvandenbossche** [comment]
@mingwandroid 

```
joris@joris-XPS-13-9350:~$ conda create -n test-mpl python=3.5 matplotlib -c defaults
Using Anaconda API: https://api.anaconda.org
Fetching package metadata ...........
Solving package specifications: ..........

Package plan for installation in environment /home/joris/miniconda3/envs/test-mpl:

The following NEW packages will be INSTALLED:

    cycler:           0.10.0-py35_0    
    dbus:             1.10.10-0        
    expat:            2.1.0-0          
    fontconfig:       2.11.1-6         
    freetype:         2.5.5-1          
    glib:             2.43.0-1         
    gst-plugins-base: 1.8.0-0          
    gstreamer:        1.8.0-0          
    icu:              54.1-0           
    jpeg:             8d-2             
    libffi:           3.2.1-0          
    libgcc:           5.2.0-0          
    libpng:           1.6.22-0         
    libxcb:           1.12-1           
    libxml2:          2.9.2-0          
    matplotlib:       1.5.3-np111py35_0
    mkl:              11.3.3-0         
    numpy:            1.11.2-py35_0    
    openssl:          1.0.2j-0         
    pip:              8.1.2-py35_0     
    pyparsing:        2.1.4-py35_0     
    pyqt:             5.6.0-py35_0     
    python:           3.5.2-0          
    python-dateutil:  2.5.3-py35_0     
    pytz:             2016.7-py35_0    
    qt:               5.6.0-1          
    readline:         6.2-2            
    setuptools:       27.2.0-py35_0    
    sip:              4.18-py35_0      
    six:              1.10.0-py35_0    
    sqlite:           3.13.0-0         
    tk:               8.5.18-0         
    wheel:            0.29.0-py35_0    
    xz:               5.2.2-0          
    zlib:             1.2.8-3          

Proceed ([y]/n)? y
...

joris@joris-XPS-13-9350:~$ source activate test-mpl
(test-mpl) joris@joris-XPS-13-9350:~$ python
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:53:06) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib.pyplot as plt
....
ImportError: No module named 'PyQt4'
>>> 
```

which is logical as the `matplotlibrc` still lists PyQt4 as the backend.

**Comment by mingwandroid** [comment]
Can you locate the `matplotlibrc` that is being used here? Please read https://github.com/ContinuumIO/anaconda-issues/issues/1068#issuecomment-255794677

**Comment by jorisvandenbossche** [comment]
So in `/home/joris/miniconda3/envs/test-mpl/lib/python3.5/site-packages/matplotlib/mpl-data`, I have `backend      : Qt4Agg` (and it is newly created env)

**Comment by mingwandroid** [comment]
Weird, can I see:
`conda list --show-channel-urls`

And also, your values for:
`sha1sum $CONDA_PREFIX/../../pkgs/matplotlib-1.5.3-np111py35_0.tar.bz2 
999f99e8c9ee28c66d325c2a8d7f2cd277900614  ..`

and:
sha1sum $CONDA_PREFIX/../../pkgs/matplotlib-1.5.3-np111py35_0/lib/python3.5/site-packages/matplotlib/mpl-data/matplotlibrc 
26b568f3cf24a297404b750a54be48eef632b2bb  ..`

.. please.

**Comment by jorisvandenbossche** [comment]
Possible reason: when this was updated in the matplotlib build, the build number was not updated (it is still 0). Therefore conda does not download the new package, but still used the package I already downloaded before end of September

**Comment by mingwandroid** [comment]
Urgh, yeah, I wonder why the build was not updated. We should make another build then.

**Comment by jorisvandenbossche** [comment]
Yep, that should solve it (I can confirm that after removing the cached matplotlib packages and again creating a fresh env, the issue is then solved)

**Comment by mingwandroid** [comment]
OK, reopening until we make a build with a bumped build number. Thanks.

**Comment by samkennerly** [comment]
For anyone who Googled this looking for a quick workaround:

#### Problem

I had the same problem @jorisvandenbossche mentioned in [this comment](https://github.com/ContinuumIO/anaconda-issues/issues/1068#issuecomment-255809756): `import matplotlib.pyplot` in a new, clean environment causes `ImportError: No module named 'PyQt4'`.

Several of my team's custom packages use `import matplotlib.pyplot`, so this glitch brought our entire Python data-science stack to a dramatic halt. Oops.

#### Workaround

As @jorisvandenbossche suggested at the top of this thread, you can pin `pyqt` to a slightly older version by running this at the command line:
```
conda install pyqt=4.11
```

#### Background

I'm using Python 3.5.2 with Anaconda 4.2.0 on Debian Linux 8.

My "new, clean environment" is a Docker container run from a custom image based on  `continuumio/anaconda3:4.2.0` plus several extra packages installed by `apt-get`, `conda`, and `pip`.

**Comment by hitvoice** [comment]
This problem occurs on my Linux server, not on my Mac. Since on a server I won't use the interactive plotting, my workaround is to use the following code
```python
import matplotlib
print(matplotlib.matplotlib_fname())
```
to get the location of the config file, and set the backend to `backend: Agg`.
Maybe it's helpful for some people who also come across `ImportError: No module named 'PyQt4'` but don't really need to call the `plt.plot()` function. The `plt.savefig()` function can be used as usual.

**Comment by aluenkinglee** [comment]
@samkennerly Thanks! it works fine!

**Comment by JeisonSosa** [comment]
Using `matplotlib-2.0.0-np111py27_0` and `pyqt-5.6.0-py27_2` the problem `ImportError: No module named 'PyQt4'` was solved updating `libxcb-1.12-0` to `libxcb-1.12-1` through

`conda install libxcb`

**Comment by boulund** [comment]
Just made a fresh install of `conda 4.3.14` (Python 3.6) on a fresh ScientifixLinux box. 
I get the `No module named 'PyQt4'` error like everyone else. Tried installing `libxcb` but that didn't work. 

Can't install `pyqt=4.11` because of no existing build (at least via conda) for Python 3.6

**Comment by abettati** [comment]
Fresh install of anaconda on ubuntu 14.04 and getting this. Is there a straight forward solution to this?

**Comment by indranilsinharoy** [comment]
Can't install `pyqt=4.11` with `conda version 4.3.13` and `Python 3.6` on `Windows 10`.

**Comment by mihirskulkarni** [comment]
Can't install `pyqt=4.11` with `conda version 4.13.0` and `Python 3.6` on Linux. Got following error:
```
UnsatisfiableError: The following specifications were found to be in conflict:
  - pyqt 4.11* -> python 2.6* -> openssl 1.0.1*
  - python 3.6*
```
`matplotlib.use('qt5agg')` also didn't work.

**Comment by amelio-vazquez-reina** [comment]
We are having the same problem @mihirskulkarni  described here. That is, we tried using:
```
backend : Qt5Agg
```
in `matplotlibrc` without luck, as well as installing `pyqt4.11` with conda on Python 3.6 on Linux. None of these solutions worked.

**Comment by ccordoba12** [comment]
Could you try:

    conda update libxcb

as @sosaj suggested above?

**Comment by brahmbhattspandan** [comment]
I am facing the same issue as @mihirskulkarni  and @amelio-vazquez-reina .  I have a fresh installation of  **Anaconda 4.3.1** and **Python 3.6.0**
Tried the following ways but **none of them seem to have worked**. 

1. `matplotlib.use("qt5agg")` 
`matplotlib.get_backend()` gives Qt5Agg

```shell
$ conda list | grep qt
pyqt                      5.6.0                    py36_2
qt                        5.6.2                         3
```
```python
In [1]: from PyQt5 import QtCore, QtGui, QtWidgets
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-43848d5bd21e> in <module>()
----> 1 from PyQt5 import QtCore, QtGui, QtWidgets

ImportError: libGL.so.1: cannot open shared object file: No such file or directory
```

2. Manually setting `matplotlibrc ` file to
`backend      : Q5Agg`

3. `conda install matplotlib --force`

4.  
```shell
$ conda install pyqt=4.11
Fetching package metadata ...........
Solving package specifications: .


UnsatisfiableError: The following specifications were found to be in conflict:
  - anaconda-navigator
  - pyqt 4.11*
  - python 3.6*
Use "conda info <package>" to see the dependencies for each package.
```

5.  `conda update libxcb`

**Comment by amelio-vazquez-reina** [comment]
Thanks @ccordoba12 Yes, I also tried `conda update libxcb` without luck. I also tried all the solutions described above, as some of the other commenters, and none of them worked.

**Comment by ccordoba12** [comment]
@amelio-vazquez-reina, there has to be something special about your Linux installation, or an error in Anaconda.

What happens when you run

```python
from PyQt5 import QtCore, QtGui, QtWidgets
```

in IPython?

Also, what's the result of

    qmake -query

in a terminal?

**Comment by amelio-vazquez-reina** [comment]
Thanks. Here's what I get:

```
$ qmake -query
QT_SYSROOT:
QT_INSTALL_PREFIX:/opt/dataxu/dxuser/anaconda3
QT_INSTALL_ARCHDATA:/opt/dataxu/dxuser/anaconda3
QT_INSTALL_DATA:/opt/dataxu/dxuser/anaconda3
QT_INSTALL_DOCS:/opt/dataxu/dxuser/anaconda3/doc
QT_INSTALL_HEADERS:/opt/dataxu/dxuser/anaconda3/include/qt
QT_INSTALL_LIBS:/opt/dataxu/dxuser/anaconda3/lib
QT_INSTALL_LIBEXECS:/opt/dataxu/dxuser/anaconda3/libexec
QT_INSTALL_BINS:/opt/dataxu/dxuser/anaconda3/bin
QT_INSTALL_TESTS:/opt/dataxu/dxuser/anaconda3/tests
QT_INSTALL_PLUGINS:/opt/dataxu/dxuser/anaconda3/plugins
QT_INSTALL_IMPORTS:/opt/dataxu/dxuser/anaconda3/imports
QT_INSTALL_QML:/opt/dataxu/dxuser/anaconda3/qml
QT_INSTALL_TRANSLATIONS:/opt/dataxu/dxuser/anaconda3/translations
QT_INSTALL_CONFIGURATION:/opt/dataxu/dxuser/anaconda3
QT_INSTALL_EXAMPLES:/opt/dataxu/dxuser/anaconda3/examples
QT_INSTALL_DEMOS:/opt/dataxu/dxuser/anaconda3/examples
QT_HOST_PREFIX:/opt/dataxu/dxuser/anaconda3
QT_HOST_DATA:/opt/dataxu/dxuser/anaconda3
QT_HOST_BINS:/opt/dataxu/dxuser/anaconda3/bin
QT_HOST_LIBS:/opt/dataxu/dxuser/anaconda3/lib
QMAKE_SPEC:linux-g++
QMAKE_XSPEC:linux-g++
QMAKE_VERSION:3.0
QT_VERSION:5.6.2
```
and

```
$ python -c 'from PyQt5 import QtCore, QtGui, QtWidgets'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: libGL.so.1: cannot open shared object file: No such file or directory
```

**Comment by ccordoba12** [comment]
So your problem (as well as @brahmbhattspandan) is the lack of `libGL.so.1`. In Debian/Ubuntu that library is provided by `libgl1-mesa-glx`. You have to find what package is needed for your distribution.

@mingwandroid, should we add a conda mesa package and make Qt 5 depend on it?

**Comment by mingwandroid** [comment]
I don't think we can just keep adding system packages as we run into missing system libraries on various linux distros as we'd just end up with the `Continuum Linux Distro`.

I would be +1 on adding some tests at install time to find missing system libraries and report them on the console instead.

**Comment by ccordoba12** [comment]
> I don't think we can just keep adding system packages

Agreed.

> I would be +1 on adding some tests at install time to find missing system libraries and report them on the console instead

Sounds like a good plan. We can do this in the PyQt5 package to make things easier (i.e. like the test above: `python -c 'from PyQt5 import QtCore, QtGui, QtWidgets'`).

**Comment by indranilsinharoy** [comment]
@ccordoba12 I'm facing the same problem (using `Python 3.6` on `Windows 10`).  Also, there is no `libxcb` port for Windows 64. What is my best option?

**Comment by ccordoba12** [comment]
@indranilsinharoy, Windows is a different story (and there's no `libxcb` there because its' not necessary). What's the output of

    qmake -query

in your case?

**Comment by indranilsinharoy** [comment]
@ccordoba12 Thank you very much for helping. Here is what I get:

```
λ qmake -query
QT_SYSROOT:
QT_INSTALL_PREFIX:C:/Users/Indranil/Anaconda3/Library
QT_INSTALL_ARCHDATA:C:/Users/Indranil/Anaconda3/Library
QT_INSTALL_DATA:C:/Users/Indranil/Anaconda3/Library
QT_INSTALL_DOCS:C:/Users/Indranil/Anaconda3/Library/doc
QT_INSTALL_HEADERS:C:/Users/Indranil/Anaconda3/Library/include/qt
QT_INSTALL_LIBS:C:/Users/Indranil/Anaconda3/Library/lib
QT_INSTALL_LIBEXECS:C:/Users/Indranil/Anaconda3/Library/bin
QT_INSTALL_BINS:C:/Users/Indranil/Anaconda3/Library/bin
QT_INSTALL_TESTS:C:/Users/Indranil/Anaconda3/Library/tests
QT_INSTALL_PLUGINS:C:/Users/Indranil/Anaconda3/Library/plugins
QT_INSTALL_IMPORTS:C:/Users/Indranil/Anaconda3/Library/imports
QT_INSTALL_QML:C:/Users/Indranil/Anaconda3/Library/qml
QT_INSTALL_TRANSLATIONS:C:/Users/Indranil/Anaconda3/Library/translations
QT_INSTALL_CONFIGURATION:
QT_INSTALL_EXAMPLES:C:/Users/Indranil/Anaconda3/Library/examples
QT_INSTALL_DEMOS:C:/Users/Indranil/Anaconda3/Library/examples
QT_HOST_PREFIX:C:/Users/Indranil/Anaconda3/Library
QT_HOST_DATA:C:/Users/Indranil/Anaconda3/Library
QT_HOST_BINS:C:/Users/Indranil/Anaconda3/Library/bin
QT_HOST_LIBS:C:/Users/Indranil/Anaconda3/Library/lib
QMAKE_SPEC:win32-msvc2015
QMAKE_XSPEC:win32-msvc2015
QMAKE_VERSION:3.0
QT_VERSION:5.6.2
```

**Comment by ccordoba12** [comment]
That seems correct. And what do you get if you run

    python -c 'from PyQt5 import QtCore, QtGui, QtWidgets'

?

**Comment by jbencina** [comment]
I have the same issue running Python 3.6.0 from Anaconda 4.3.1 installed today on Ubuntu 16.01

```
me@main-compute:~/notebooks$ python -c 'from PyQt5 import QtCore, QtGui, QtWidgets'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: libGL.so.1: cannot open shared object file: No such file or directory

```

**Comment by ccordoba12** [comment]
@jbencina, please install the package I mentioned above: `libgl1-mesa-glx`

**Comment by David-OConnor** [comment]
Checking back in: This mess is one of the reason I switched to vanilla Python; hope it gets sorted eventually!

**Comment by ccordoba12** [comment]
Sorry, but I don't think there's a mess here. As far as we can see, there are only missing libraries or packages on some installations that prevent Qt5 to work.

As @mingwandroid correctly pointed out, we can't keep adding libraries to make Qt5 depend on because we would end up creating the *Continuum Linux Distro*.

**Comment by jbencina** [comment]
@ccordoba12 Yes, I spoke too soon and missed that in the thread. Installing `libgl1-mesa-glx` on Ubuntu indeed corrects the problem and allows for rendering within a Jupyter Notebook now.

**Comment by ccordoba12** [comment]
@jbencina, great!! Thanks for letting us know about it :-)

**Comment by indranilsinharoy** [comment]
@ccordoba12 . Thank you very much, once again. I am very grateful for your help. You asked me to run `python -c 'from PyQt5 import QtCore, QtGui, QtWidgets'` . I did, and I didn't get any errors (or any other message/output). I guess that the modules imported properly.

**Comment by ccordoba12** [comment]
@indranilsinharoy, what happens if you run `ipython` in a terminal, and then run these commands on it?

```
In [1]: %matplotlib qt5

In [2]: import matplotlib.pyplot as plt

In [3]: plt.plot(range(10))
```

**Comment by indranilsinharoy** [comment]
@ccordoba12 Thank you very much. The problem is resolved now. I changed the backend to `Qt5agg` in the `matplotlibrc` file, and that worked (earlier when I made the same change I had also, by mistake, changed `backend.qt4`  to `PyQt5` which caused the errors).  
Thanks a lot.

**Comment by mihirskulkarni** [comment]
I installed miniconda2 from scratch again on a linux system (NASA Pleiades). After entering `from PyQt5 import QtCore, QtGui, QtWidgets` in IPython, I get the following error similar to @tskisner : 
`ImportError: /usr/lib64/libxcb-xlib.so.0: undefined symbol: _xcb_unlock_io`

`libxcb` is up-to-date. However the workaround of installing `pyqt=4.11` which worked for @tskisner, didn't work for me - gave the same error.

**Comment by ccordoba12** [comment]
@mihirskulkarni, it seems you have a very old X stack installed in your system:

* https://lists.freedesktop.org/archives/xcb/2009-April/004489.html
* https://cgit.freedesktop.org/xcb/libxcb/commit/?id=d989656cde2ee7a4a66b2065209ef389495f3452

Again, not our fault. We support up to CentOS 5 systems with their latest updates installed.

**Comment by mihirskulkarni** [comment]
@ccordoba12 I don't have much hope then. Thanks for your help.

**Comment by ccordoba12** [comment]
@mihirskulkarni, what Python version are you using?

**Comment by mihirskulkarni** [comment]
`Python 2.7.13` installed using `Miniconda2-latest-Linux-x86_64.sh`.

**Comment by ccordoba12** [comment]
So why are not you able to use pyqt 4?

**Comment by mihirskulkarni** [comment]
For the default installation, I got `ImportError: No module named PyQt4`. I then installed PyQt4 with `conda install pyqt=4.11`. When I do `from matplotlib import pyplot` after that from IPython, I get `ImportError: /usr/lib64/libxcb-xlib.so.0: undefined symbol: _xcb_unlock_io`.

**Comment by ccordoba12** [comment]
Ok, so it seems the problem affects Qt4 packages too. Sorry for not being able to give you more help.

**Comment by simonm3** [comment]
I had this issue using ubuntu 14.04 docker image. Installing libgl1-mesa-glx solved it thanks.

I guess docker images normally don't have any need for installing GUI packages. However data scientists often run jupyter notebooks using matplotlib on a server using docker. If anaconda is aimed at data scientists it would be useful for it to add this package if it is required to get matplotlib to work.

**Comment by mingwandroid** [comment]
We cannot add system libraries like this for if we did, our software would not run the best libgl for the hardware on a given end-user's computer.

**Comment by simonm3** [comment]
We are only talking about the case where matplotlib fails because there is
no libgl installed such as a docker container. So you could conditionally
install one. Most people who install anaconda are going to use matplotlib
for which a libgl is a requirement so it would be the "batteries included"
solution.

**Comment by mingwandroid** [comment]
Conda had no support for conditionally installing things. It would be nice if a post install script was provided to warn and advise though I think.

**Comment by simonm3** [comment]
A warning would be useful. I tried all the other solutions listed for this and none of them worked until I installed libgl.

**Comment by mingwandroid** [comment]
OK, I will raise this with my team.

**Comment by ccordoba12** [comment]
@mingwandroid, I have a simple idea on how to solve this. I'll create a PR in our internal repo to start the discussion about it.

**Comment by simonm3** [comment]
Actually I  am not sure how the error arose. It happened on an ubuntu docker container running on AWS centos host. However I have the same container running on windows fine. It is possible I upgraded something which caused the failure but I cannot be certain.

**Comment by madhavij** [comment]
I was getting the following issue:
```

$ python
Python 3.5.2 |Anaconda custom (64-bit)| (default, Jul  2 2016, 17:53:06) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
>>> import matplotlib.pyplot as plt
/data/anaconda3/lib/python3.5/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
/data/anaconda3/lib/python3.5/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py", line 114, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "/data/anaconda3/lib/python3.5/site-packages/matplotlib/backends/__init__.py", line 32, in pylab_setup
    globals(),locals(),[backend_name],0)
  File "/data/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt4agg.py", line 18, in <module>
    from .backend_qt5agg import FigureCanvasQTAggBase as _FigureCanvasQTAggBase
  File "/data/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt5agg.py", line 15, in <module>
    from .backend_qt5 import QtCore
  File "/data/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt5.py", line 31, in <module>
    from .qt_compat import QtCore, QtGui, QtWidgets, _getSaveFileName, __version__
  File "/data/anaconda3/lib/python3.5/site-packages/matplotlib/backends/qt_compat.py", line 124, in <module>
    from PyQt4 import QtCore, QtGui
ImportError: libXdmcp.so.6: cannot open shared object file: No such file or directory

```
I did a yum install for the libXdmcp package and that helped.

**Comment by ccordoba12** [comment]
@madhavij, that issue was solved in a new build of our Qt5 package. A simple

    conda update qt

should solve it too.

**Comment by schekroud** [comment]
>For the default installation, I got ImportError: No module named PyQt4. I then installed PyQt4 with conda install pyqt=4.11. When I do from matplotlib import pyplot after that from IPython, I get ImportError: /usr/lib64/libxcb-xlib.so.0: undefined symbol: _xcb_unlock_io.<

has this been sorted? I just updated anaconda because having problems with it, but now can't open spyder at all, giving me the following error:

>For the default installation, I got ImportError: No module named PyQt4. I then installed PyQt4 with conda install pyqt=4.11. When I do from matplotlib import pyplot after that from IPython, I get ImportError: /usr/lib64/libxcb-xlib.so.0: undefined symbol: _xcb_unlock_io.<

**Comment by schekroud** [comment]
>For the default installation, I got ImportError: No module named PyQt4. I then installed PyQt4 with conda install pyqt=4.11. When I do from matplotlib import pyplot after that from IPython, I get ImportError: /usr/lib64/libxcb-xlib.so.0: undefined symbol: _xcb_unlock_io. <

has this been sorted? I just updated anaconda because having problems with it, but now can't open spyder at all, giving me the following error:

>Traceback (most recent call last):
  File "/home/sammirc/anaconda2/bin/spyder", line 6, in <module>
    sys.exit(spyder.app.start.main())
  File "/home/sammirc/anaconda2/lib/python2.7/site-packages/spyder/app/start.py", line 103, in main
    from spyder.app import mainwindow
  File "/home/sammirc/anaconda2/lib/python2.7/site-packages/spyder/app/mainwindow.py", line 78, in <module>
    from qtpy.compat import from_qvariant
  File "/home/sammirc/anaconda2/lib/python2.7/site-packages/qtpy/compat.py", line 15, in <module>
    from .QtWidgets import QFileDialog
  File "/home/sammirc/anaconda2/lib/python2.7/site-packages/qtpy/QtWidgets.py", line 24, in <module>
    from PyQt4.QtGui import *
ImportError: /usr/lib64/libxcb-xlib.so.0: undefined symbol: _xcb_unlock_io <

**Comment by ccordoba12** [comment]
@schekroud, it seems you're using very old distro. What distro are you using?

**Comment by braaannigan** [comment]
Hi all
This problem has persisted for me for weeks now.  I had a working version with pyqt4.11 and matplotlib1.5 and updated to matplotlib2.0.1.  Since then the only workaround that has worked has been to use the Agg backend, which limits me from some necessary interactive operations.
I've tried without success:
downloading a new version of anaconda (Anaconda2-4.3.1-Linux-x86_64)
modifying the .matplotlibrc file to qt4 and qt5 backends
pinning pyqt back to version 4.1.1.

Does anyone have additional suggestions?

**Comment by schekroud** [comment]
The distro wasn't too old (a year old, maybe). Something seemed to have gone wrong with anaconda.

@braaannigan I sorted this by completely stripping anaconda from my computer, including the hidden files folders that you'd see if you went to your terminal and tried `ls -a`. I deleted all remnants of anaconda, including backed up config files, and then reinstalled anaconda and all the packages I used. Now, not an issue, and i'm back to work. Not sure why it happened in the first place given the correct modules were installed..

**Comment by kokalz** [comment]
braaannigan, have you tried:
`conda update qt?`
this one worked for me

**Comment by braaannigan** [comment]
Thanks @schekroud, I've given that a go, but it hasn't worked.  What version of anaconda did you reinstall?

Hi @kokalz , I've tried updating it to 5.6 and putting it back to 4.11 numerous times.  What do you get from conda list?

**Comment by kokalz** [comment]
@braaannigan just checked and while this solved an issue for python3 I seem to still be getting this issue in python2. Which python are you using?

**Comment by reMAJ** [comment]
@kokalz 
I am using python 2.7.13. Updated all conda packages qt as well. I still get the issue. Any solution please?
thanks

**Comment by braaannigan** [comment]
@reMAJ 
Hi - I've given up trying to solve this.  On the system where it happens, I'm able to work with everything as it should be apart from ipython, where I'm stuck using version 3.0.0. If I try to update ipython then everything stops working.  I'm just trying to pretend that I'm a hipster for using out-of-date software and hoping that a clean build in a couple of months will solve it.   Sorry I can't be of more use.

**Comment by ankeshkhemani** [comment]
`sudo apt-get install libsm6 libxrender1 gcc g++ libqt4-dev
`
This is what fixed the issue for me

**Comment by eflanigan00** [comment]
Also having this problem with Anaconda 4.4.0 on RHEL 7

**Comment by santiagocasas** [comment]
None of these solutions have worked for me! I am on Linux and conda 4.3.25.

**Comment by eflanigan00** [comment]
@santiagocasas  try removing all your ~/.matplotlib, ~/.ipython and other configuration directories.  My problem was resolved by resetting all of these.  Of course save any custom config files to another location before deleting them.

**Comment by bennywij** [comment]
Hi — we have a conda installation on a headless Centos 7 box and ran into the same problem. We solved it (thanks in part to this thread and the standard Google/StackOverflow searching) by:

`yum install mesa-libGL`

and, by turning interactive mode off for matplotlib per https://github.com/ContinuumIO/anaconda-issues/issues/1215

**Comment by esvhd** [comment]
I encountered this on my MacBook Pro, running Anaconda with Python 3.6.1, matplotlib 2.0.2.

My `matplotlibrc` was using `qt4agg` as the backend, after changing that to `qt5agg`, the was fixed.

**Comment by hzhu6666** [comment]
python 2.7 on Linux...this has been bothering me for quite a while.....yum install mesa-libGL is the way to go

**Comment by kelhallaoui** [comment]
I am having the same error. I have uninstalled and reinstalled Anaconda twice with no lock. I do not have non-ASCII characters in my path. 

If i try to change the backend same problem occurs

```
In [1]: import matplotlib

In [2]: matplotlib.get_backend()
Out[2]: 'Qt5Agg'
In [3]: import matplotlib.pyplot as plt

In [4]: This application failed to start because it could not find or load the Qt platform plugin "windows"
in "".
```

Available platform plugins are: minimal, offscreen, windows.

Reinstalling the application may fix this problem.

By packages are listed below:

`Python 3.6.3 |Anaconda custom (64-bit)| (default, Nov  8 2017, 15:10:56) [MSC v.1900 64 bit (AMD64)] on win32`

```
# packages in environment at C:\Users\eee\Anaconda3:
#
_ipyw_jlab_nb_ext_conf    0.1.0            py36he6757f0_0
alabaster                 0.7.10           py36hcd07829_0
anaconda                  custom           py36h363777c_0
anaconda-client           1.6.5            py36hd36550c_0
anaconda-navigator        1.6.9            py36hc720852_0
anaconda-project          0.8.0            py36h8b3bf89_0
asn1crypto                0.22.0           py36h8e79faa_1
astroid                   1.5.3            py36h9d85297_0
astropy                   2.0.2            py36h06391c4_4
babel                     2.5.0            py36h35444c1_0
backports                 1.0              py36h81696a8_1
backports.functools_lru_cache 1.4                      py36_1    conda-forge
backports.shutil_get_terminal_size 1.0.0            py36h79ab834_2
beautifulsoup4            4.6.0            py36hd4cc5e8_1
bitarray                  0.8.1            py36h6af124b_0
bkcharts                  0.2              py36h7e685f7_0
blaze                     0.11.3           py36h8a29ca5_0
bleach                    2.0.0            py36h0a7e3d6_0
bokeh                     0.12.10          py36h0be3b39_0
boto                      2.48.0           py36h1a776d2_1
bottleneck                1.2.1            py36hd119dfa_0
bzip2                     1.0.6                haa5b126_2
ca-certificates           2017.08.26           h94faf87_0
cachecontrol              0.12.3           py36hfe50d7b_0
certifi                   2017.7.27.1      py36h043bc9e_0
cffi                      1.10.0           py36hae3d1b5_1
chardet                   3.0.4            py36h420ce6e_1
click                     6.7              py36hec8c647_0
cloudpickle               0.4.0            py36h639d8dc_0
clyent                    1.2.2            py36hb10d595_1
colorama                  0.3.9            py36h029ae33_0
comtypes                  1.1.2            py36heb9b3d1_0
conda                     4.3.30           py36h7e176b0_0
conda-build               3.0.30           py36he296fd2_0
conda-env                 2.6.0                h36134e3_1
conda-verify              2.0.0            py36h065de53_0
console_shortcut          0.1.1                h6bb2dd7_3
contextlib2               0.5.5            py36he5d52c0_0
cryptography              2.0.3            py36h123decb_1
curl                      7.55.1               h3b839b5_4
cycler                    0.10.0           py36h009560c_0
cython                    0.26.1           py36h18049ac_0
cytoolz                   0.8.2            py36h547e66e_0
dask                      0.16.0           py36hea4f2f0_0
dask-core                 0.16.0           py36h5683b08_0
datashape                 0.5.4            py36h5770b85_0
decorator                 4.1.2            py36he63a57b_0
distlib                   0.2.5            py36h51371be_0
distributed               1.20.0           py36h2fcf393_0
docutils                  0.14             py36h6012d8f_0
entrypoints               0.2.3            py36hfd66bb0_2
et_xmlfile                1.0.1            py36h3d2d736_0
fastcache                 1.0.2            py36hffdae1b_0
filelock                  2.0.12           py36hd7ddd41_0
flask                     0.12.2           py36h98b5e8f_0
flask-cors                3.0.3            py36h8a3855d_0
freetype                  2.8                  h51f8f2c_1
get_terminal_size         1.0.0                h38e98db_0
gevent                    1.2.2            py36h342a76c_0
glob2                     0.5              py36h11cc1bd_1
greenlet                  0.4.12           py36ha00ad21_0
h5py                      2.7.0            py36hfbe0a52_1
hdf5                      1.10.1               h98b8871_1
heapdict                  1.0.0            py36h21fa5f4_0
html5lib                  0.999999999      py36ha09b1f3_0
icc_rt                    2017.0.4             h97af966_0
icu                       58.2                 ha66f8fd_1
idna                      2.6              py36h148d497_1
imageio                   2.2.0            py36had6c2d2_0
imagesize                 0.7.1            py36he29f638_0
intel-openmp              2018.0.0             hcd89f80_7
ipykernel                 4.6.1            py36hbb77b34_0
ipython                   6.1.0            py36h236ecc8_1
ipython_genutils          0.2.0            py36h3c5d0ee_0
ipywidgets                7.0.0            py36h2e74ada_0
isort                     4.2.15           py36h6198cc5_0
itsdangerous              0.24             py36hb6c5a24_1
jdcal                     1.3              py36h64a5255_0
jedi                      0.10.2           py36hed927a0_0
jinja2                    2.9.6            py36h10aa3a0_1
jpeg                      9b                   hb83a4c4_2
jsonschema                2.6.0            py36h7636477_0
jupyter                   1.0.0            py36h422fd7e_2
jupyter_client            5.1.0            py36h9902a9a_0
jupyter_console           5.2.0            py36h6d89b47_1
jupyter_core              4.3.0            py36h511e818_0
jupyterlab                0.27.0           py36h34cc53b_2
jupyterlab_launcher       0.4.0            py36h22c3ccf_0
krb5                      1.14.2               h9d72080_5
lazy-object-proxy         1.3.1            py36hd1c21d2_0
libiconv                  1.15                 h1df5818_7
libpng                    1.6.32           vc14h5163883_3  [vc14]
libssh2                   1.8.0            vc14hcf584a9_2  [vc14]
libtiff                   4.0.8           vc14h04e2a1e_10  [vc14]
libxml2                   2.9.4            vc14h8fd0f11_5  [vc14]
libxslt                   1.1.29           vc14hf85b8d4_5  [vc14]
llvmlite                  0.20.0                   py36_0
locket                    0.2.0            py36hfed976d_1
lockfile                  0.12.2           py36h0468280_0
lxml                      4.1.0            py36h0dcd83c_0
lzo                       2.10             vc14h0a64fa6_1  [vc14]
markupsafe                1.0              py36h0e26971_1
matplotlib                2.1.0            py36h11b4b9c_0
mccabe                    0.6.1            py36hb41005a_1
menuinst                  1.4.10           py36h42196fb_0
mistune                   0.8.1            py36h007b88b_0
mkl                       2018.0.0             h36b65af_4
mkl-service               1.1.2            py36h57e144c_4
mpmath                    0.19             py36he326802_2
msgpack-python            0.4.8            py36h58b1e9d_0
multipledispatch          0.4.9            py36he44c36e_0
navigator-updater         0.1.0            py36h8a7b86b_0
nbconvert                 5.3.1            py36h8dc0fde_0
nbformat                  4.4.0            py36h3a5bc1b_0
networkx                  2.0              py36hff991e3_0
nibabel                   2.2.0                    py36_0    conda-forge
nltk                      3.2.4            py36hd0e0a39_0
nose                      1.3.7            py36h1c3779e_2
notebook                  5.2.1            py36h4fb2ca6_0
numba                     0.35.0             np113py36_10
numexpr                   2.6.2            py36h514de0f_2
numpy                     1.13.3           py36ha320f96_0
numpydoc                  0.7.0            py36ha25429e_0
odo                       0.5.1            py36h7560279_0
olefile                   0.44             py36h0a7bdd2_0
openpyxl                  2.4.8            py36hf3b77f6_1
openssl                   1.0.2m               h093b818_1
packaging                 16.8             py36ha0986f6_1
pandas                    0.21.0           py36he09d4dd_1
pandoc                    1.19.2.1             hb2460c7_1
pandocfilters             1.4.2            py36h3ef6317_1
partd                     0.3.8            py36hc8e763b_0
path.py                   10.3.1           py36h3dd8b46_0
pathlib2                  2.3.0            py36h7bfb78b_0
patsy                     0.4.1            py36h42cefec_0
pep8                      1.7.0            py36h0f3d67a_0
pickleshare               0.7.4            py36h9de030f_0
pillow                    4.2.1            py36hdb25ab2_0
pip                       9.0.1            py36h226ae91_4
pkginfo                   1.4.1            py36hb0f9cfa_1
ply                       3.10             py36h1211beb_0
progress                  1.3              py36hbeca8d3_0
prompt_toolkit            1.0.15           py36h60b8f86_0
psutil                    5.4.0            py36h4e662fb_0
py                        1.4.34           py36ha4aca3a_1
pycodestyle               2.3.1            py36h7cc55cd_0
pycosat                   0.6.3            py36h413d8a4_0
pycparser                 2.18             py36hd053e01_1
pycrypto                  2.6.1            py36he68e6e2_1
pycurl                    7.43.0           py36h086bf4c_3
pydicom                   0.9.9                    py36_0    conda-forge
pyflakes                  1.6.0            py36h0b975d6_0
pygments                  2.2.0            py36hb010967_0
pylint                    1.7.4            py36ha4e6ded_0
pyodbc                    4.0.17           py36h0006bc2_0
pyopenssl                 17.2.0           py36h15ca2fc_0
pyparsing                 2.2.0            py36h785a196_1
pyqt                      5.6.0            py36hb5ed885_5
pysocks                   1.6.7            py36h698d350_1
pytables                  3.4.2            py36h71138e3_2
pytest                    3.2.1            py36h753b05e_1
python                    3.6.3                h3b118a2_4
python-dateutil           2.6.1            py36h509ddcb_1
pytz                      2017.2           py36h05d413f_1
pywavelets                0.5.2            py36hc649158_0
pywin32                   221              py36h9c10281_0
pyyaml                    3.12             py36h1d1928f_1
pyzmq                     16.0.2           py36h38c27d9_2
qt                        5.6.2           vc14h6f8c307_12  [vc14]
qtawesome                 0.4.4            py36h5aa48f6_0
qtconsole                 4.3.1            py36h99a29a9_0
qtpy                      1.3.1            py36hb8717c5_0
requests                  2.18.4           py36h4371aae_1
rope                      0.10.5           py36hcaf5641_0
ruamel_yaml               0.11.14          py36h9b16331_2
scikit-image              0.13.0           py36h6dffa3f_1
scikit-learn              0.19.1           py36h53aea1b_0
scipy                     1.0.0            py36h1260518_0
seaborn                   0.8.0            py36h62cb67c_0
setuptools                36.5.0           py36h65f9e6e_0
simplegeneric             0.8.1            py36heab741f_0
singledispatch            3.4.0.3          py36h17d0c80_0
sip                       4.18.1           py36h9c25514_2
six                       1.11.0           py36h4db2310_1
snowballstemmer           1.2.1            py36h763602f_0
sortedcollections         0.5.3            py36hbefa0ab_0
sortedcontainers          1.5.7            py36ha90ac20_0
sphinx                    1.6.3            py36h9bb690b_0
sphinxcontrib             1.0              py36hbbac3d2_1
sphinxcontrib-websupport  1.0.1            py36hb5e5916_1
spyder                    3.2.4            py36h8845eaa_0
sqlalchemy                1.1.13           py36h5948d12_0
sqlite                    3.20.1               h9eeafa9_2
statsmodels               0.8.0            py36h6189b4c_0
sympy                     1.1.1            py36h96708e0_0
tblib                     1.3.2            py36h30f5020_0
testpath                  0.3.1            py36h2698cfe_0
tk                        8.6.7                hcb92d03_3
toolz                     0.8.2            py36he152a52_0
tornado                   4.5.2            py36h57f6048_0
traitlets                 4.3.2            py36h096827d_0
typing                    3.6.2            py36hb035bda_0
unicodecsv                0.14.1           py36h6450c06_0
urllib3                   1.22             py36h276f60a_0
vc                        14                   h2379b0c_2
vs2015_runtime            14.0.25123           hd4c4e62_2
wcwidth                   0.1.7            py36h3d5aa90_0
webencodings              0.5.1            py36h67c50ae_1
werkzeug                  0.12.2           py36h866a736_0
wheel                     0.29.0           py36h6ce6cde_1
widgetsnbextension        3.0.2            py36h364476f_1
win_inet_pton             1.0.1            py36he67d7fd_1
win_unicode_console       0.5              py36hcdbd4b5_0
wincertstore              0.2              py36h7fe50ca_0
wrapt                     1.10.11          py36he5f5981_0
xlrd                      1.1.0            py36h1cb58dc_1
xlsxwriter                1.0.2            py36hf723b7d_0
xlwings                   0.11.4           py36hd3cf94d_0
xlwt                      1.3.0            py36h1a4751e_0
yaml                      0.1.7                hc54c509_2
zict                      0.1.3            py36h2d8e73e_0
zlib                      1.2.11               h8395fce_2

```

**Comment by nothing2wang** [comment]
Hi, I have the same problem  use  spyder  by Anaconda

>>>import matplotlib.pyplot as plt


Traceback (most recent call last):

  File "<ipython-input-2-eff513f636fd>", line 1, in <module>
    import matplotlib.pyplot as plt

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py", line 115, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/backends/__init__.py", line 32, in pylab_setup
    globals(),locals(),[backend_name],0)

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt4agg.py", line 18, in <module>
    from .backend_qt5agg import FigureCanvasQTAggBase as _FigureCanvasQTAggBase

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt5agg.py", line 16, in <module>
    from .backend_qt5 import QtCore

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/backends/backend_qt5.py", line 26, in <module>
    import matplotlib.backends.qt_editor.figureoptions as figureoptions

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/backends/qt_editor/figureoptions.py", line 20, in <module>
    import matplotlib.backends.qt_editor.formlayout as formlayout

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/backends/qt_editor/formlayout.py", line 56, in <module>
    from matplotlib.backends.qt_compat import QtGui, QtWidgets, QtCore

  File "/home/qingsong/anaconda3/lib/python3.5/site-packages/matplotlib/backends/qt_compat.py", line 141, in <module>
    from PyQt4 import QtCore, QtGui

ImportError: No module named 'PyQt4'
___________________________________________________
>>>python --version
Python 3.5.4 :: Anaconda custom (64-bit)

But  I use python by the terminal
>>>python
>>> import matplotlib.pyplot as plt

That works fine! why?

**Comment by nothing2wang** [comment]
@ccordoba12  Hi, I have the same problem use spyder by Anaconda

>>import matplotlib.pyplot as plt

import matplotlib.pyplot as plt
Traceback (most recent call last):

  File "<ipython-input-4-eff513f636fd>", line 1, in <module>
    import matplotlib.pyplot as plt

  File "/home/qingsong/anaconda3/lib/python3.6/site-packages/matplotlib/pyplot.py", line 116, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()

  File "/home/qingsong/anaconda3/lib/python3.6/site-packages/matplotlib/backends/__init__.py", line 60, in pylab_setup
    [backend_name], 0)

  File "/home/qingsong/anaconda3/lib/python3.6/site-packages/matplotlib/backends/backend_qt4agg.py", line 10, in <module>
    from .backend_qt4 import (

  File "/home/qingsong/anaconda3/lib/python3.6/site-packages/matplotlib/backends/backend_qt4.py", line 18, in <module>
    from .qt_compat import QtCore, QtWidgets, _getSaveFileName, __version__

  File "/home/qingsong/anaconda3/lib/python3.6/site-packages/matplotlib/backends/qt_compat.py", line 150, in <module>
    from PyQt4 import QtCore, QtGui

ModuleNotFoundError: No module named 'PyQt4'
-----------------------------------------------------------------------------------------
$python --version
Python 3.6.3 :: Anaconda custom (64-bit)
-------------------------------------------------------------------------------------
But I use python by the terminal

$python
>>>import matplotlib.pyplot as plt
>>>

That works fine!  why ? 
how to fix this problem on spyder ?

thanks a lot !
